streaming/base/converters/dataframe_to_mds.py

```diff
@@ -0,0 +1,326 @@
+
+import collections
+import json
+import logging
+import os
+import shutil
+from argparse import ArgumentParser, Namespace
+from collections.abc import Iterable
+from typing import Any, Callable, Dict, Iterable, Optional, Tuple, Union
+
+import pandas as pd
+from pyspark import TaskContext
+from pyspark.sql import SparkSession
+from pyspark.sql.dataframe import DataFrame
+from pyspark.sql.types import (ArrayType, BinaryType, BooleanType, ByteType, DateType,
+                               DayTimeIntervalType, DecimalType, DoubleType, FloatType,
+                               IntegerType, LongType, MapType, ShortType, StringType, StructField,
+                               StructType, TimestampNTZType, TimestampType)
+
+from streaming import MDSWriter
+from streaming.base.format.index import get_index_basename
+from streaming.base.format.mds.encodings import _encodings
+from streaming.base.storage.upload import CloudUploader
+
+logger = logging.getLogger(__name__)
+
+
+def is_iterable(obj: Any) -> bool:
+    """Check if obj is iterable."""
+    return issubclass(type(obj), Iterable)
+
+
+def infer_dataframe_schema(dataframe: DataFrame,
+                           user_defined_cols: Optional[Dict[str, Any]] = None) -> Optional[Dict]:
+    """Takes a pyspark dataframe and retrives the schema information, constructing a dictionary."""
+    dtype_mapping = {
+        ByteType: 'bytes',
+        ShortType: 'uint64',
+        IntegerType: 'int',
+        LongType: 'int64',
+        FloatType: 'float32',
+        DoubleType: 'float64',
+        DecimalType: None,
+        StringType: 'str',
+        BinaryType: None,
+        BooleanType: None,
+        TimestampType: None,
+        TimestampNTZType: None,
+        DateType: None,
+        DayTimeIntervalType: None,
+        ArrayType: None,
+        MapType: None,
+        StructType: None,
+        StructField: None
+    }
+
+    def map_spark_dtype(spark_data_type: Any):
+        for k, v in dtype_mapping.items():
+            if isinstance(spark_data_type, k) and v is not None:
+                return v
+        raise ValueError(f'{spark_data_type} is not supported by MDSwrite')
+
+    if user_defined_cols is not None:  # user has provided schema, we just check if mds supports the dtype
+        mds_supported_dtypes = list(dtype_mapping.values())
+        for k, v in user_defined_cols.items():
+            if k not in dataframe.columns:
+                raise ValueError(f'{k} is not a column of input dataframe')
+            if v not in mds_supported_dtypes:
+                raise ValueError(f'{v} is not supported by MDSwrite')
+        return None
+
+    schema = dataframe.schema
+    schema_dict = {}
+
+    for field in schema:
+        dtype = map_spark_dtype(field.dataType)
+        if dtype in _encodings:
+            schema_dict[field.name] = dtype
+        else:
+            print(_encodings)
+
+    return schema_dict
+
+
+def do_merge_index(partitions: Iterable,
```

Comment:
Wondering, can we unify a method for spark and non-spark users ? We can have a non-spark user method `merge_index()` which gets called inside `do_merge_index()` ? Your thoughts ?

Reply:
Not sure I understand. Can you expand it a little bit?

Reply:
As the method does not have `spark.sql.DataFrame` but `pandas.DataFrame` dependencies, the question seems to be whether the code can be moved to a generic place. That would allow reusability through multiple multi-processing approaches.

Reply:
@XiaohanZhangCMU Currently, there is utility method to merge index.json for non-spark user dataset conversion usage. I am thinking if we can create a generic function for non-spark users (where they do not use dataframeToMDS() method) and the spark users.

Reply:
@kurosch @karan6181 It would be great if it could be in a module script. IE. `python -m streaming.merge_indices index_01.json ... index_99.json index_merged.json`.